Home Work

KamalSyu opened the pull request.

KamalSyu committed:
Вариант домашнего задания №1.Есть много моментов, которые нуждаются в доработке. В процессе улучшения.

KamalSyu committed:
Вариант домашнего задания №1.Есть много моментов, которые нуждаются в доработке. В процессе улучшения.

KamalSyu committed:
Домашнее задание №2.

KamalSyu committed:
home work3

KamalSyu committed:
Домашнее задание №2.

KamalSyu committed:
Домашнее задание №3 .

KamalSyu committed:
Домашнее задание №3 (последняя версия).

Liberate520 commented on src/ru/gb/family_tree/family_tree/FamilyTree.java:
```diff
@@ -0,0 +1,114 @@
+        return sb.toString();
+    }
+
+    public void sortByName() { humanList.sort((Comparator<? super Human>) new FamilyTreeComparatorByName());}
```
но классы, объект которого вы создаете в качестве компаратора пустые и соответственно никакой сортировки здесь не будет. Более того, я думаю, что такой проект даже не соберется, так как приведение типа будет невозможно. Если объект изначально не является компаратором, то вы никак не сможете его таковым сделать

KamalSyu committed:
Домашнее задание №4

KamalSyu committed:
Новые изменения, чтобы код работал

KamalSyu committed:
New changes

KamalSyu committed:
Домашнее задание №5

KamalSyu committed:
New structure

KamalSyu committed:
Домашнее задание №5
Структура проекта
ru.gb2.family_tree7
│
├── Main.java
│
├── view
│   ├── FamilyTreeView.java
│   └── ConsoleFamilyTreeView.java
│
├── presenter
│   └── FamilyTreePresenter.java
│
├── model
│   ├── writer
│   │   ├── Writable.java
│   │   └── FileHandler.java
│   │
│   ├── human
│   │   ├── Human.java
│   │   └── Gender.java
│   │
│   └── family_tree
│       ├── FamilyTree.java
│       ├── FamilyTreeComparatorByBirthDate.java
│       ├── FamilyTreeComparatorByName.java
│       ├── FamilyTreeIterator.java
│       └── TreeNode.java
### Описание классов и интерфейсов

#### 1. `Main.java`
Класс, в котором начинается выполнение программы. Здесь создается экземпляр `FamilyTreePresenter`, который инициализирует приложение и запускает главный цикл взаимодействия с пользователем.

#### 2. `view`

- **Interface `FamilyTreeView`**
- **Методы:**
- `void showMembers(List<Human> members);`
- Показывает список членов семьи.
- `void showMessage(String message);`
- Выводит сообщения пользователю.

- **Класс `ConsoleFamilyTreeView`**
- Реализует интерфейс `FamilyTreeView`.
- **Методы:**
- `void showMembers(List<Human> members)` - выводит на консоль список членов семьи.
- `void showMessage(String message)` - выводит сообщения пользователю в консоли.

#### 3. `presenter`

- **Класс `FamilyTreePresenter`**
- Связывает модель (данные) с представлением (интерфейс).
- Содержит методы для добавления членов семьи и отображения списка.
- **Методы:**
- `void addMember(Human member)` - добавляет нового члена семьи.
- `void showMembers()` - отображает всех членов семьи через `FamilyTreeView`.

#### 4. `model`

- **writer**

- **Interface `Writable`**
- Интерфейс для классов, которые могут записывать свои данные.
- **Методы:**
- `void writeToFile(String filename);`

- **Класс `FileHandler`**
- Реализует интерфейс `Writable`.
- **Методы:**
- `void writeToFile(String filename)` - реализует логику записи данных в файл.

- **human**

- **Класс `Human`**
- Представляет члена семьи.
- **Поля:**
- `String name`
- `Gender gender`
- `LocalDate dateOfBirth`
- **Методы:**
- `int getAge()`
- `String toString()`

- **Enum `Gender`**
- Перечисление, представляющее пол.
- **Значения:**
- `MALE`, `FEMALE`

- **family_tree**

- **Класс `FamilyTree`**
- Представляет семейное дерево.
- **Поля:**
- `List<Human> members`
- **Методы:**
- `void addMember(Human member)`
- `List<Human> getMembers()`

- **Класс `FamilyTreeComparatorByBirthDate`**
- Компаратор для сортировки членов семьи по дате рождения.
- **Методы:**
- `int compare(Human h1, Human h2)`

- **Класс `FamilyTreeComparatorByName`**
- Компаратор для сортировки членов семьи по имени.
- **Методы:**
- `int compare(Human h1, Human h2)`

- **Класс `FamilyTreeIterator`**
- Итератор для обхода членов семьи в семейном дереве.
- **Методы:**
- `boolean hasNext()`
- `Human next()`

- **Interface `TreeNode`**
- Интерфейс для узлов дерева (опционально, если вы хотите создать более сложную структуру дерева).
- **Методы:**
- `List<TreeNode> getChildren()`
- `void addChild(TreeNode child)`

### Заключение

Эта структура проекта позволяет четко организовать код, разделив его на логические компоненты. Каждый класс и интерфейс имеет свою определенную ответственность, что улучшает читаемость и облегчает поддержку и расширение проекта в будущем.

Liberate520 commented on src/ru/gb2/family_tree7/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,55 @@
+//        }
+//    }
+private final FamilyTreeView view;
+    private final List<Human> members; // Список членов семьи
```
Нужно создать класс Сервис. А класс Сервис будет работать с классом семейного дерева. И вот класс Семейного дерева будет работать со списком людей, а не презентер

Liberate520 commented on src/ru/gb2/family_tree7/view/ConsoleFamilyTreeView.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+
+        return new Human(name, gender, birthDate, role);
```
view не должно создавать объекты модели (людей)

Liberate520 commented on src/ru/gb2/family_tree7/model/human/Human.java:
```diff
@@ -0,0 +1,114 @@
+    private Gender gender;
+    private List<Human> parents = new ArrayList<>(); // Список родителей
+    private List<Human> children = new ArrayList<>(); // Список детей
+    private String role; // Роль в семье (мама, папа, ребенок)
```
понятие ребенок или родитель являются относительными. Один и тот же человек может быть одновременно и родителем и ребенком

Liberate520 commented on src/ru/gb2/family_tree7/model/human/Human.java:
```diff
@@ -0,0 +1,114 @@
+        this.role = role;
+    }
+
+    public Human(String name, Gender gender, int birthYear, String role) {
+    }
+
+    public Human(String name, Gender gender, LocalDate birthDate, String role) {
+    }
```
В чем смысл передавать какие то аргументы если с этими аргументами ничего не происходит?

KamalSyu committed:
Домашнее задание №56

KamalSyu committed:
Проект семейного древа

KamalSyu committed:
save new status

KamalSyu committed:
Домашнее задание 6

Liberate520 commented on src/ru/gb2/family_tree7/presenter/Presenter.java:
```diff
@@ -0,0 +1,35 @@
+        service = new Service();
+    }
+
+    public void addHuman(String name, int age) {
```
Слишком мало полей для создания человека

Liberate520 commented on src/ru/gb2/family_tree7/model/human/Human.java:
```diff
@@ -0,0 +1,43 @@
+
+public class Human implements Comparable<Human>, GroupItem {
+    private int id;
+    private String name;
+    private int age;
```
слишком мало полей. Мы обсуждали, что вместо age должна быть дата рождения и смерти. Проект предусматривает создание родственных связей, значит должны быть списки детей и родителей

Liberate520 commented on src/ru/gb2/family_tree7/presenter/Presenter.java:
```diff
@@ -0,0 +1,35 @@
+    public void sortByName() {
+        service.sortByName();
+        getHumanListInfo();
+    }
```
Добавить методы сохранения и загрузки

Liberate520 replied:
Методы создания родственной связи

KamalSyu committed:
Весь проект семейного древа

Liberate520 commented on src/ru/gb2/family_tree7/presenter/Presenter.java:
```diff
@@ -0,0 +1,85 @@
+public class Presenter {
+    private View view;
+    private Service service;
+    private static int idCounter = 0; // Генерация ID
```
В Сервисе аналогичная переменная. Где то точно лишняя

Liberate520 commented on src/ru/gb2/family_tree7/presenter/Presenter.java:
```diff
@@ -0,0 +1,85 @@
+
+    public void addHuman(String name, String s, Gender gender, LocalDate birthDate, LocalDate deathDate, Human father, Human mother) {
+        // Создаем новый объект Human
+        Human newHuman = new Human(getNextId(), name, gender, birthDate, deathDate); // Генерация ID
```
Создание объектов модели это работа модели, а не презентера

Liberate520 commented on src/ru/gb2/family_tree7/presenter/Presenter.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public void getHumanListInfo() {
+        List<Human> humans = service.getHumanList(); // Предположим, что у вас есть метод, который возвращает список людей
+        String info = service.getHumanInfo(); // Обновите метод, чтобы выводил информацию о всех людях
+        view.printAnswer(info);
```
Опять же зачем первое действие, если оно не используется?

Liberate520 commented on src/ru/gb2/family_tree7/model/service/Service.java:
```diff
@@ -0,0 +1,110 @@
+        Human human = new Human(getNextId(), name, gender, birthDate, deathDate);
+        return human;
+    }
+
```
Так и не реализованы методы сохранения
